### ISSUE TITLE
Explore: move linear_model mini-batch learning into Rust (true per-sample SGD)

### ISSUE THREAD

**Issue by MaxHalford** [online-ml/river#1912]
# Explore: move `linear_model` mini-batch learning into Rust (true per-sample SGD)

## Background: `learn_many` is not equivalent to looping `learn_one`

Today `linear_model`'s `GLM.learn_many` (used by `LinearRegression` /
`LogisticRegression`) does **mini-batch gradient descent**: it computes the mean
gradient over the whole batch and applies a **single** weight update
(`rust_src`-free, in `river/linear_model/base.py`):

```python
# _eval_gradient_many
gradient = np.einsum("ij,i->j", X.values, loss_gradient) / len(X)  # mean over the batch
...
# _fit -> one optimizer.step per batch
```

Looping `learn_one` over the same rows instead does **N sequential updates**,
where each row's prediction reflects the updates from all preceding rows. So the
two APIs are *not* numerically equivalent. The divergence grows with batch size
(see numbers below).

For comparison, scikit-learn's streaming linear models (`SGDClassifier` /
`SGDRegressor.partial_fit`) do the per-sample thing: they loop over samples
one at a time, updating after each, inside compiled Cython. Vowpal Wabbit does
the same in C++. So "push the loop into a compiled language" is a well-trodden
design, not a novelty — and, as the cross-check below shows, with a constant
learning rate and no penalty sklearn's `SGDClassifier` computes *exactly* the
same update River does.

## Proposal (for a future, dedicated release)

Add a Rust fast-path for `learn_many` that performs **true per-sample SGD** —
making `learn_many` numerically identical to looping `learn_one`, but without
crossing the Python/FFI boundary per row. This would:

- remove a surprising semantic gap between `learn_one` and `learn_many`,
- match the online-learning "use each sample promptly" philosophy of the library,
- be much faster than the Python `learn_one` loop.

This issue documents a **prototype** + benchmark so we know what a real
implementation would entail. It is intentionally *not* meant to ship as-is.

## Prototype

A dependency-free Rust kernel (`rust_src/linear.rs`, exposed as
`river._river_rust.linear.sgd_log_fit_many`) restricted to the **default
`LogisticRegression` config**: `SGD` with a constant learning rate, `Log` loss,
constant intercept LR, optional L2 (no L1). It receives the batch as a
row-major little-endian f64 byte buffer, the current dense weight vector and
intercept, and returns the updated `(weights, intercept)` after a full
per-sample pass. The Log-loss gradient (including the `|z|>18` stability
branches) and gradient clipping are mirrored exactly from
`optim/losses.py` / `linear_model/base.py`.

Benchmark: `benchmarks/linear_minibatch_rust_prototype.py` (synthetic dense data,
n=50k, p=40), comparing five ways to consume the data in equal batches:

1. River `learn_one` loop — online ground truth
2. River `learn_many` (current, averaged-gradient)
3. River Rust per-sample (this prototype)
4. **sklearn `SGDClassifier.partial_fit`** — `log_loss`, `learning_rate='constant'`,
   `eta0=0.01`, `penalty=None` (same algorithm, per-sample SGD in Cython)
5. **Vowpal Wabbit** — `--sgd --loss_function logistic --learning_rate 0.01`,
   single streaming pass (per-sample SGD in C++)

### Results

**Equivalence** — three independent implementations of constant-LR per-sample
logistic SGD (River `learn_one`, our Rust kernel, sklearn `SGDClassifier`) all
agree to machine epsilon. This validates the prototype *and* shows there is a
single well-defined "true online SGD" answer they all target:

| comparison | max\|Δw\| | \|Δintercept\| |
|---|---|---|
| **rust vs River learn_one** | **6.7e-16** | 5.6e-17 |
| **rust vs sklearn** | **4.4e-16** | 0.0 |
| **sklearn vs River learn_one** | **4.4e-16** | 5.6e-17 |
| River learn_many vs learn_one (batch 64) | 1.18 | 0.25 |

(VW is excluded from the numeric check — its internal LR schedule, per-feature
adaptive scaling and normalization differ, so it is a throughput reference only,
not a bit-for-bit match.)

The current `learn_many` diverges from `learn_one`, and that divergence **grows
with batch size** (more averaging ⇒ further from online SGD):

| batch | learn_many vs learn_one, max\|Δw\| |
|---|---|
| 16 | 0.69 |
| 64 | 1.18 |
| 256 | 1.52 |
| 1024 | 1.67 |
| 4096 | 1.72 |

**Throughput** (ex/s, n=50k, p=40; in-process libraries are model-only, VW is
end-to-end incl. its text parsing). All per-sample columns are the *same
algorithm*; only the implementation differs.

| batch | River learn_one | River learn_many (current) | **River rust per-sample** | sklearn partial_fit | VW (end-to-end) |
|---|---|---|---|---|---|
| 64   | 56k | 1.71M | **3.82M** | 0.66M | ~0.47M |
| 256  | 51k | 6.20M | **7.56M** | 1.91M | ~0.49M |
| 1024 | 53k | 12.5M | 11.4M | 4.43M | ~0.56M |
| 4096 | 52k | 23.8M | 11.9M | — | — |

Takeaways:

- Rust per-sample is **~25–230× faster than the Python `learn_one` loop** while
  being numerically identical to it.
- Rust per-sample is **~2.5–5.8× faster than sklearn's Cython per-sample loop**
  and ~8× faster than VW's end-to-end pass (caveat: VW includes text parsing and
  uses a different, adaptive update). So a Rust per-sample River would be very
  competitive with the established compiled online learners.
- vs. River's current `learn_many`: Rust **wins on small batches** (per-batch
  Python overhead dominates the averaged path) and **loses on large batches**.
  Crossover is ~batch 1024. The reason is fundamental: per-sample updates have a
  sequential dependency and cannot be vectorized, so the Rust kernel saturates
  ~12M ex/s, whereas `learn_many`'s single `X @ w` matmul keeps scaling via BLAS.

So the win from this move is **correctness/equivalence + a big speedup over
one-by-one and over sklearn/VW**, *not* unconditionally beating River's own
current batched path on raw throughput at large batch sizes.

## What a real implementation requires (the hard part)

The prototype's speed comes entirely from not calling back into Python per row.
But River's GLM update is highly configurable, and each knob is a Python object
touched per step in the online path:

- **~15 optimizers** with internal state (`optim/`: SGD, Adam, FTRLProximal,
  AdaGrad, RMSProp, Momentum, Nesterov, AdaDelta, AdaMax, AMSGrad, Nadam,
  AdaBound, Newton, …) via `optimizer.step` / `optimizer.look_ahead`.
- **loss functions** (`optim/losses.py`) — regression + binary.
- **learning-rate schedulers** (`optim/schedulers.py`) and a separate intercept LR.
- **initializers** (`optim/initializers.py`).
- **L1 cumulative-penalty** state machine (`_update_weights`), which currently
  isn't even correctly wired for the batch path (it iterates `x.items()`,
  fine for a dict, not a DataFrame).

To keep the speed, these would have to be reimplemented in Rust — which is a
large surface area. The pragmatic scope (and what sklearn does) is to support a
**fixed enum** of optimizer/loss/scheduler/penalty combinations in Rust and fall
back to the existing Python path for anything exotic.

Open questions for the real version:

- Zero-copy input via `PyReadonlyArray2<f64>` (adds the `numpy` Rust crate,
  version-locked to PyO3) instead of the prototype's `tobytes()` buffer.
- How to keep `optimizer.n_iterations` and adaptive-LR state consistent across
  the Rust loop (the prototype assumes constant LRs).
- Sparse features: a dense matrix is natural for `learn_many`, but River models
  are often used with sparse/hashed features.
- API: do we silently reroute `learn_many` (behaviour change — results will move
  for everyone) or expose it opt-in? Given the divergence numbers, changing
  `learn_many` semantics is a breaking change worth a major version.

## Suggested plan

1. (this prototype) — measure feasibility & equivalence ✅
2. Dedicated release: Rust kernels for the common configs (SGD + a few
   losses + constant/inv-scaling schedules + L2), opt-in or behind a version flag.
3. Decide whether per-sample becomes the default `learn_many` semantics (breaking).

cc — prototype code on branch, not for merge: `rust_src/linear.rs`,
`benchmarks/linear_minibatch_rust_prototype.py`.
